server/vmap/trillian_map_server.go

```diff
@@ -74,24 +74,33 @@ func (t *TrillianMapServer) GetLeaves(ctx context.Context, req *trillian.GetMapL
 
 	smtReader := merkle.NewSparseMerkleTreeReader(req.Revision, hasher, tx)
 
-	leaves, err := tx.Get(ctx, req.Revision, req.Index)
-	if err != nil {
-		return nil, err
-	}
-	glog.Infof("%v: wanted %v leaves, found %v", mapID, len(req.Index), len(leaves))
+	// TODO: we need to return one response per request.
+	// Get only returns items for things that exist.
 
-	resp := &trillian.GetMapLeavesResponse{
-		MapLeafInclusion: make([]*trillian.MapLeafInclusion, len(leaves)),
-	}
-	for i, leaf := range leaves {
-		proof, err := smtReader.InclusionProof(ctx, req.Revision, leaf.Index)
+	inclusions := make([]*trillian.MapLeafInclusion, 0, len(req.Index))
+	for i, index := range req.Index {
+		proof, err := smtReader.InclusionProof(ctx, req.Revision, index)
 		if err != nil {
 			return nil, err
 		}
-		// Copy the leaf from the iterator, which gets overwritten
-		value := leaf
+
+		// Create placeholder leaf for non-existant leaves.
+		leaf := &trillian.MerkleLeaf{
+			Index: index,
+		}
+
+		// Fetch the leaf if it exists.
+		leafs, err := tx.Get(ctx, req.Revision, [][]byte{index})
+		if err != nil {
+			return nil, err
+		}
+		for l := range leafs {
+			leaf = l
+			break
+		}
+
```

Comment:
I've given this a bit more thought and I think the corner cases we're handling here actually merit error responses, as they're likely to stem from bugs / mistaken requests.

E.g.:
* `len(req.Index) == 0` -> `InvalidArgument` (no sense in asking for zero leaves)
* `len(index) == 0` -> `InvalidArgument` (no sense in asking for a leaf with an empty index)
* leaf not in storage -> `NotFound` (unambiguous response, unlike a empty-but-valid proof)

(See [status.Errorf](https://godoc.org/google.golang.org/grpc/status#Errorf).)